cpp/src/arrow/flight/sql/server_test.cc

```diff
@@ -153,12 +168,46 @@ class TestFlightSqlServer : public ::testing::Environment {
 
   void TearDown() override {
     sql_client.reset();
-    server->Stop();
-    server.reset();
+
+    ASSERT_OK(server->Shutdown());
+    server_thread->join();
+    server_thread.reset();
+  }
+
+ private:
+  std::string unix_sock;
+  std::shared_ptr<arrow::flight::sql::example::SQLiteFlightSqlServer> server;
+  std::unique_ptr<std::thread> server_thread;
+  std::condition_variable server_ready_cv;
+  std::mutex server_ready_m;
+
+  Status RunServerInternal() {
+    arrow::flight::Location location;
+    ARROW_CHECK_OK(arrow::flight::Location::ForGrpcUnix(unix_sock, &location));
+    arrow::flight::FlightServerOptions options(location);
+
+    ARROW_ASSIGN_OR_RAISE(server,
+                          arrow::flight::sql::example::SQLiteFlightSqlServer::Create())
+
+    ARROW_CHECK_OK(server->Init(options));
+    // Exit with a clean error code (0) on SIGTERM
+    ARROW_CHECK_OK(server->SetShutdownOnSignals({SIGTERM}));
+
+    server_ready_cv.notify_all();
+    ARROW_CHECK_OK(server->Serve());
+
+    return arrow::Status::OK();
+  }
+
+  static void RunServer(TestFlightSqlServer* server) {
+    arrow::Status st = server->RunServerInternal();
+    if (!st.ok()) {
+      std::cerr << st << std::endl;
+    }
```

Comment:
```suggestion
    ASSERT_OK(server->RunServerInternal());
```

Reply:
Though ideally it would abort the test if setup fails.

Reply:
Actually, above we're using `ARROW_CHECK_OK` so maybe we can avoid having `RunServer` and `RunServerInternal`.